File: 14-cache/python-sdk/cache_control.py
"""
Cost Katana Cache: Cache Control Operations (Python)

Clear, warmup, export, and import cache.

Run: python 18-cache/python-sdk/cache_control.py
"""

import os
import requests
from typing import List, Optional

API_BASE = "https://cost-katana-backend.store/api"
API_KEY = os.getenv("COST_KATANA_API_KEY")

def clear_cache(model: Optional[str] = None, older_than: Optional[int] = None):
    """Clear cache"""
    
    params = {}
    if model:
        params["model"] = model
    if older_than:
        params["olderThan"] = older_than
    
    response = requests.delete(
        f"{API_BASE}/cache/clear",
        params=params,
        headers={"Authorization": f"Bearer {API_KEY}"}
    )
    
    response.raise_for_status()
    data = response.json()["data"]
    
    print("✅ Cache cleared:")
    print(f"  Entries cleared: {data['entriesCleared']}")
    print(f"  Bytes freed: {data['bytesFreed'] / 1024 / 1024:.2f} MB")
    
    return data

def warmup_cache(prompts: List[str], model: str = "gpt-4", ttl: int = 86400):
    """Warmup cache with common queries"""
    
    response = requests.post(
        f"{API_BASE}/cache/warmup",
        json={"prompts": prompts, "model": model, "ttl": ttl},
        headers={
            "Authorization": f"Bearer {API_KEY}",
            "Content-Type": "application/json"
        }
    )
    
    response.raise_for_status()
    data = response.json()["data"]
    
    print("✅ Cache warmup complete:")
    print(f"  Cached: {data['cached']}")
    print(f"  Failed: {data['failed']}")
    print(f"  Total Cost: ${data['totalCost']:.4f}")
    
    return data

def export_cache(format: str = "json"):
    """Export cache data"""
    
    response = requests.get(
        f"{API_BASE}/cache/export",
        params={"format": format},
        headers={"Authorization": f"Bearer {API_KEY}"}
    )
    
    response.raise_for_status()
    print(f"✅ Cache exported ({format})")
    
    return response.json()

def main():
    if not API_KEY:
        print("❌ COST_KATANA_API_KEY required")
        return
    
    print("🥷 Cache Control Operations\n")
    
    try:
        # 1. Warmup cache
        print("1️⃣ Warming up cache...")
        warmup_cache([
            "What is AI?",
            "Explain machine learning",
            "What is deep learning?"
        ])
        
        # 2. Export cache
        print("\n2️⃣ Exporting cache...")
        export_cache()
        
        # 3. Clear old entries
        print("\n3️⃣ Clearing old entries...")
        clear_cache(older_than=86400)
        
        print("\n✅ Cache control operations complete!")
        
    except requests.exceptions.HTTPError as e:
        print(f"❌ Error: {e.response.json() if e.response else e}")

if __name__ == "__main__":
    main()

File: 14-cache/npm-package/cache-control.ts
/**
 * Cost Katana Cache: Cache Control Operations
 * 
 * Clear, warmup, export, and import cache.
 * 
 * Run: npx ts-node 18-cache/npm-package/cache-control.ts
 */

import axios from 'axios';

const API_BASE = 'https://cost-katana-backend.store/api'\;
const API_KEY = process.env.COST_KATANA_API_KEY;

async function clearCache(filters?: { model?: string; olderThan?: number }) {
  try {
    const params = new URLSearchParams();
    if (filters?.model) params.append('model', filters.model);
    if (filters?.olderThan) params.append('olderThan', filters.olderThan.toString());

    const response = await axios.delete(
      `${API_BASE}/cache/clear?${params.toString()}`,
      {
        headers: { 'Authorization': `Bearer ${API_KEY}` }
      }
    );

    console.log('✅ Cache cleared:');
    console.log(`  Entries cleared: ${response.data.data.entriesCleared}`);
    console.log(`  Bytes freed: ${(response.data.data.bytesFreed / 1024 / 1024).toFixed(2)} MB`);

    return response.data.data;
  } catch (error: any) {
    console.error('❌ Failed:', error.response?.data || error.message);
    throw error;
  }
}

async function warmupCache(prompts: string[], model: string = 'gpt-4', ttl: number = 86400) {
  try {
    const response = await axios.post(
      `${API_BASE}/cache/warmup`,
      { prompts, model, ttl },
      {
        headers: {
          'Authorization': `Bearer ${API_KEY}`,
          'Content-Type': 'application/json'
        }
      }
    );

    console.log('✅ Cache warmup complete:');
    console.log(`  Cached: ${response.data.data.cached}`);
    console.log(`  Failed: ${response.data.data.failed}`);
    console.log(`  Total Cost: $${response.data.data.totalCost.toFixed(4)}`);

    return response.data.data;
  } catch (error: any) {
    console.error('❌ Failed:', error.response?.data || error.message);
    throw error;
  }
}

async function exportCache(format: 'json' | 'csv' = 'json') {
  try {
    const response = await axios.get(
      `${API_BASE}/cache/export?format=${format}`,
      {
        headers: { 'Authorization': `Bearer ${API_KEY}` }
      }
    );

    console.log(`✅ Cache exported (${format})`);
    console.log(`  Entries: ${response.data.entries?.length || 0}`);

    return response.data;
  } catch (error: any) {
    console.error('❌ Failed:', error.response?.data || error.message);
    throw error;
  }
}

async function importCache(entries: any[]) {
  try {
    const response = await axios.post(
      `${API_BASE}/cache/import`,
      { entries },
      {
        headers: {
          'Authorization': `Bearer ${API_KEY}`,
          'Content-Type': 'application/json'
        }
      }
    );

    console.log('✅ Cache import complete:');
    console.log(`  Imported: ${response.data.data.imported}`);
    console.log(`  Failed: ${response.data.data.failed}`);
    console.log(`  Duplicates: ${response.data.data.duplicates}`);

    return response.data.data;
  } catch (error: any) {
    console.error('❌ Failed:', error.response?.data || error.message);
    throw error;
  }
}

async function main() {
  if (!API_KEY) {
    console.error('❌ COST_KATANA_API_KEY required');
    process.exit(1);
  }

  console.log('🥷 Cache Control Operations\n');

  try {
    // 1. Warmup cache
    console.log('1️⃣ Warming up cache...');
    await warmupCache([
      'What is AI?',
      'Explain machine learning',
      'What is deep learning?'
    ]);

    // 2. Export cache
    console.log('\n2️⃣ Exporting cache...');
    await exportCache();

    // 3. Clear old cache entries
    console.log('\n3️⃣ Clearing old cache entries...');
    await clearCache({ olderThan: 86400 }); // Older than 24 hours

    console.log('\n✅ Cache control operations complete!');
  } catch (error) {
    console.error('\n❌ Failed:', error);
    process.exit(1);
  }
}

if (require.main === module) {
  main();
}

export { clearCache, warmupCache, exportCache, importCache };
